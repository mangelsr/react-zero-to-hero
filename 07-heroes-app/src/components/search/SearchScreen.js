import React, { useMemo } from 'react';
import queryString from "query-string";
import { useLocation } from 'react-router-dom';

import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../ui/heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';


export const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    
    const [ formValues, handleInputChange ] = useForm({ search: q });
    
    const { search } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])
    
    const submitForm = (e) => {
        e.preventDefault();
        history.push(`?q=${search}`);
    };

    return (
        <div>
            <h1>Search</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search form</h4>
                    <form onSubmit={submitForm}>
                        <input
                            value={search}
                            onChange={handleInputChange}
                            autoComplete="off"
                            name="search"
                            type="text"
                            placeholder="Find your hero"
                            className="form-control" />
                        <button
                            type="submit"
                            className="btn btn-outline-primary btn-block mt-4">
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <br/>
                    { (q === '') &&
                        <div className="alert alert-info">Search a hero</div>
                    }
                    { (q !== '' && heroesFiltered.length === 0) &&
                        <div className="alert alert-warning">No hero found for {q}</div>
                    }
                    { heroesFiltered.map(hero => <HeroCard key={hero.id} {...hero} />)}
                </div>
            </div>
        </div>
    )
}
