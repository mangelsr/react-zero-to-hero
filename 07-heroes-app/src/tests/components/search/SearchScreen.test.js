import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('Tests over <SearchScreen /> component', () => {
    
    test('should redner the component with the default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').exists()).toBe(true);
    });

    test('should redner the component with existing hero search', () => {
        const heroName = 'batman';
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${heroName}`]}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe(heroName);
    });

    test('should redner the component with non existing hero search', () => {
        const heroName = 'batman-not-existing';
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${heroName}`]}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-warning').exists()).toBe(true);
    });

    test('should call push of history', () => {
        const inputText = 'batman';
        const histoyMock = {
            push: jest.fn()
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search`]}>
                <Route path="/search" component={() => <SearchScreen history={histoyMock} /> } />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change', {
            target: { name: 'search', value: inputText }
        });
        wrapper.find('form').simulate('submit', { preventDefault(){} });
        expect(histoyMock.push).toHaveBeenCalledWith(`?q=${inputText}`);
    });

});