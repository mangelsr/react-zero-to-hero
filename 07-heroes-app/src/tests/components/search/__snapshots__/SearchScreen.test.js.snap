// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <SearchScreen /> component should redner the component with existing hero search 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary btn-block mt-4"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <br />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                className="btn btn-outline-primary"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests over <SearchScreen /> component should redner the component with non existing hero search 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman-not-existing"
        />
        <button
          className="btn btn-outline-primary btn-block mt-4"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <br />
      <div
        className="alert alert-warning"
      >
        No hero found for 
        batman-not-existing
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests over <SearchScreen /> component should redner the component with the default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary btn-block mt-4"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <br />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
